src/mlpack/methods/decision_tree/decision_tree_impl.hpp

```diff
@@ -690,24 +690,24 @@ double DecisionTree<FitnessFunction,
     // Get the number of children we will have.
     size_t numChildren = 0;
     if (datasetInfo.Type(bestDim) == data::Datatype::categorical)
-      numChildren = CategoricalSplit::NumChildren(classProbabilities, *this);
+      numChildren = CategoricalSplit::NumChildren(classProbabilities[0], *this);
     else
-      numChildren = NumericSplit::NumChildren(classProbabilities, *this);
+      numChildren = NumericSplit::NumChildren(classProbabilities[0], *this);
```

Comment:
I think we should handle it but in a separate PR. Earlier I was thinking that I will deal with things later, but now I am realising that Decision Trees need to be perfect because the performance of many other methods depends directly on their performance.

It might mess up a little with the timeline, but I think that this needs proper attention. What do you think?

Reply:
I agree that this could make a big difference.  Why don't we handle the prefix-sum issue first, then see what the performance is?  I have spent some time thinking about it, and I suspect that the change actually won't be all that hard to implement.  Also I do think it is okay to delay the timeline a bit for this issue.  My assumption is that it would probably take maybe a week to implement the change, debug it, and do some quick benchmarking---let me know what you're thinking, you probably have a better picture than I do. :) @gmanlan any thoughts on that?

Reply:
I agree with you both about this being important for the PR completeness. Think about it, without prefix-sum and proper handling of memory used by nodes we may struggle to achieve the goals proposed for XGB (that being, comparable performance as benchmarked against reference implementations). That said, I would recommend addressing these two because it's more like an investment for the right foundation - if you implement XGB first and then realize that it blows up memory or runs quite slowly, then we will need to come back to this anyway so maybe sooner rather than later?

Reply:
@rcurtin @gmanlan, Okay. So, let me first wrap up the remaining tests and handle the basic suggestions from the review. Then I will work on this. First I will benchmark our current implementation. I think UCI energy dataset will be good enough for this. 

I don't know if you remember, but I mentioned earlier that I was unable to figure out mathematically how to do prefix-sum in weighted scenario. Maybe you can help with that?

Reply:
> I don't know if you remember, but I mentioned earlier that I was unable to figure out mathematically how to do prefix-sum in weighted scenario. Maybe you can help with that?

@rcurtin @gmanlan I cracked this 😃. Check this out https://drive.google.com/file/d/1OH4DTL9F-FGLvC76i2TyA5rkfaS878Za/view?usp=sharing. Please let me know if my 

Earlier when I tried to solve this problem, I was expanding the summation and then trying to simplify it. That method was pretty complicated and didn't give any result. But keeping the summation as it is, the solution became pretty neat and easy. Now, I guess for MSEGain, we have reduced complexity from O(n^2) to O(n).🚀

Reply:
Yes, I believe your derivation is right.  Nice work!